lib/mauth/request_and_response.rb

```diff
@@ -23,6 +39,67 @@ def string_to_sign(more_attributes)
       Digest::SHA512.hexdigest(string)
     end
 
+    # the string to sign for V2 protocol will be (where LF is line feed character)
+    # for requests:
+    #   string_to_sign =
+    #     http_verb + <LF> +
+    #     resource_url_path (no host, port or query string; first "/" is included) + <LF> +
+    #     request_body_digest + <LF> +
+    #     app_uuid + <LF> +
+    #     current_seconds_since_epoch + <LF> +
+    #     encoded_query_params
+    #
+    # for responses:
+    #   string_to_sign =
+    #     status_code_string + <LF> +
+    #     response_body_digest + <LF> +
+    #     app_uuid + <LF> +
+    #     current_seconds_since_epoch
+    def string_to_sign_v2(more_attributes)
+      attributes_for_signing = self.attributes_for_signing.merge(more_attributes)
+
+      # lazy instantiation of body digest to avoid hashing request bodies
+      # three times because we call string to sign three times.
+      if attributes_for_signing[:body]
+        attributes_for_signing[:body_digest] ||= Digest::SHA512.hexdigest(attributes_for_signing[:body].to_s)
+      end
+      attributes_for_signing[:encoded_query_params] = encode_query_string(attributes_for_signing[:query_string].to_s)
+
+      missing_attributes = self.class::SIGNATURE_COMPONENTS_V2.reject do |key|
+        attributes_for_signing.dig(key)
+      end
+
+      missing_attributes.delete(:body_digest) # body may be omitted
+      if missing_attributes.any?
+        raise(UnableToSignError, "Missing required attributes to sign: #{missing_attributes.inspect}\non object to sign: #{inspect}")
+      end
+
+      string = self.class::SIGNATURE_COMPONENTS_V2.map do |k|
+        attributes_for_signing[k].to_s.force_encoding('UTF-8')
+      end.join("\n")
+      Digest::SHA512.hexdigest(string)
+    end
+
+    # sorts query string parameters by codepoint, uri encodes keys and values,
+    # and rejoins parameters into a query string
+    def encode_query_string(q_string)
+      q_string.split('&').sort.map do |part|
+        k, e, v = part.partition('=')
+        uri_escape(k) + e + uri_escape(v)
+      end.join('&')
+    end
+
+    # percent encodes special characters, preserving character encoding
+    # identical to CGI.escape except it encodes spaces as %20%
+    # QUESTION should just use CGI.escape then gsub?
```

Comment:
Seems worth writing a simple `Benchmark` to answer the question (not to be included in the PR)